src/coreclr/jit/gentree.cpp

```diff
@@ -18022,6 +18022,18 @@ CORINFO_CLASS_HANDLE Compiler::gtGetClassHandle(GenTree* tree, bool* pIsExact, b
                         }
                     }
                 }
+                else if (tree->TypeIs(TYP_REF) && base->IsCnsIntOrI())
+                {
+                    if (base->IsIconHandle(GTF_ICON_CONST_PTR, GTF_ICON_STATIC_HDL) &&
+                        (base->AsIntCon()->gtFieldSeq != nullptr))
+                    {
+                        CORINFO_FIELD_HANDLE fldHandle = base->AsIntCon()->gtFieldSeq->GetFieldHandle();
+                        if (fldHandle != nullptr)
+                        {
+                            objClass = gtGetFieldClassHandle(fldHandle, pIsExact, pIsNonNull);
+                        }
+                    }
+                }
```

Comment:
`tree->TypeIs(TYP_REF)` looks redundant (checked at the top of the function).

`base->IsCnsIntOrI` look like it can be replaced with `IsIconHandle(GTF_ICON_CONST_PTR, GTF_ICON_STATIC_HDL)` (since that checks for `CNS_INT` already).

`fldHandle` will never be `nullptr`.

There needs to be a check for that `fieldSeq->GetOffset() == base->IconValue()`.